ext/web/16_image_data.js

```diff
@@ -0,0 +1,174 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+import * as webidl from "ext:deno_webidl/00_webidl.js";
+import DOMException from "ext:deno_web/01_dom_exception.js";
+const primordials = globalThis.__bootstrap.primordials;
+const {
+  TypeError,
+} = primordials;
+
+const PredefinedColorSpace = [
+  "srgb",
+  "display-p3",
+];
+
+class ImageData {
+  /** @type {number} */
+  #width;
+  /** @type {height} */
+  #height;
+  /** @type {Uint8Array} */
+  #data;
+  /** @type {'srgb' | 'display-p3'} */
+  #colorSpace;
+
+  constructor() {
+    webidl.requiredArguments(
+      arguments.length,
+      2,
+      'Failed to construct "ImageData"',
+    );
+
+    const [arg0, arg1, arg2, arg3] = arguments;
+    let sourceWidth;
+    let sourceHeight;
+    let data;
+    let settings;
+
+    // Overload: new ImageData(data, sw [, sh [, settings ] ])
+    if (typeof arg0 === "object") {
+      data = arg0;
+      sourceWidth = typeof arg1 !== "undefined"
```

Comment:
all of these type conversions and validations should be done using the webidl converters where applicable